### ISSUE TITLE
Match-Making Algorithm: Score-Based

### ISSUE THREAD

**Issue by alyhasansakr** [EdgeLab-FHDO/Edge-Diagnostic-Platform#63]
[Aly Hasan Sakr - Master Thesis.pdf](https://github.com/EdgeLab-FHDO/Edge-Diagnostic-Platform/files/5579443/Aly.Hasan.Sakr.-.Master.Thesis.pdf) mentions a score-based match-making algorithm and shows some results, we wish to implement the same algorithm in our platform and run the same match-making experiment to get similar results.

We already have a [baseline implementation](https://github.com/EdgeLab-FHDO/Edge-Diagnostic-Platform/tree/master/src/main/java/InfrastructureManager/MatchMaking) that you need to extend.

Write a Scenario file to test different cases without any applications.

After #35 is finished, you will need to also test the algorithm with this application (may need minor modifications).

Don't forget automatic testing.

**Comment by zero212** [comment]
I need to find a way to update the node's stat. sending another POST for node registration will just create another duplicate (with the modified line but the old one still there)

**Comment by alyhasansakr** [comment]
this is where you need to change [addNode](https://github.com/EdgeLab-FHDO/Edge-Diagnostic-Platform/blob/f22201f27b4ec709a7cbcc9448593420628dc75e/src/main/java/InfrastructureManager/Master.java#L182)

**Comment by zero212** [comment]
Rather than modifying the addNode function. I made another one called [updateNode](https://github.com/EdgeLab-FHDO/Edge-Diagnostic-Platform/blob/2400b6759a3316a8207b7716f650be98bd3e8ff8/src/main/java/InfrastructureManager/Master.java#L195), issue here is I am not able to pass the `command` to [out](https://github.com/EdgeLab-FHDO/Edge-Diagnostic-Platform/blob/2400b6759a3316a8207b7716f650be98bd3e8ff8/src/main/java/InfrastructureManager/MatchMaking/MatchMaker.java#L106) function here. I did modify the [restInput](https://github.com/EdgeLab-FHDO/Edge-Diagnostic-Platform/blob/2400b6759a3316a8207b7716f650be98bd3e8ff8/src/main/java/InfrastructureManager/Rest/RestInput.java#L48) and [restRouter](https://github.com/EdgeLab-FHDO/Edge-Diagnostic-Platform/blob/2400b6759a3316a8207b7716f650be98bd3e8ff8/src/main/java/InfrastructureManager/Rest/RestRouter.java#L19) here though. 

Here's the console screenshot that prove it 
![image](https://user-images.githubusercontent.com/31235907/100947279-a25c5200-3505-11eb-9fe0-9fbccac7468b.png)

Any suggestion/help would be much much many many appreciated

**Comment by alyhasansakr** [comment]
I believe you need to add that to the configuration as well, just like [this](https://github.com/EdgeLab-FHDO/Edge-Diagnostic-Platform/blob/2400b6759a3316a8207b7716f650be98bd3e8ff8/src/main/resources/Configuration.json#L75)

**Comment by juan-castrillon** [comment]
@zero212  Exactly as Aly said. You need to add the `update_node` to the commands that the rest input sends to the match maker. Specifically in here:
```json
{
      "in" : "rest_in",
      "out" : "match_m",
      "commands" : {
        "register_client $client_as_json_string" : "matchMaker register_client $client_as_json_string",
        "register_node $node_as_json_string" : "matchMaker register_node $node_as_json_string",
        "assign_client $client_id" : "matchMaker assign_client $client_id"
      }
}
```

**Comment by zero212** [comment]
I know we still in progress on defining and naming things. Just gonna leave things that annoyed me a bit here such as 
The 2 lists here in Master and in matchMaker basically work the same. But they have different variable names. 🤦 
![image](https://user-images.githubusercontent.com/31235907/101050780-040fd100-3585-11eb-9f99-b3b238610306.png)

![image](https://user-images.githubusercontent.com/31235907/101050881-273a8080-3585-11eb-9a85-91c0545f5093.png)

**Comment by alyhasansakr** [comment]
You are free to make any changes you want in your branch even if it fundamentally changes other parts of the code.

**Comment by zero212** [comment]
Basically done with SBMM, here's is Postman json body for input.

Node register
localhost:4567/node/register
localhost:4567/node/update
```
{
    "id" : "node1",
    "ipAddress": "68.131.232.215 : 30968",
    "connected" : true,
    "resource" : 100,
    "network" : 100,
    "location" : 55
}
```


Client register 
localhost:4567/client/register
localhost:4567/client/update
```
{
    "id" : "client2",
    "reqNetwork" : 5,
    "reqResource" :10,
    "location" : 54
}
```

Match making
localhost:4567/client/assign/client2

Client disconnect. message field can be changed to `job_done` when disconnecting without problem
localhost:4567/client/disconnect
```
{
    "id" : "client2",
    "message" : "job_failed"
}
```

**Comment by alyhasansakr** [comment]
Great, that is almost it, but I can't say for sure unless we have a repeatable test, to so we should make a few test scenarios (instead of POST requests). This way we will be able to better analyze and test our implementations.
Remember that we need it first to function the same as the old one, then we improve it.

**Comment by zero212** [comment]
The automatic test going to take time. I may need a hand or 2 here :)) But let me see what I can do first. Curl command and github CI or gradle is nothing that I am familiar with.

**Comment by zero212** [comment]
Node to self so I don't have to look up the search history. 
- Postman do have a tool called newman to assist this automated testing https://blog.postman.com/writing-tests-in-postman//

**Comment by alyhasansakr** [comment]
We are going to implement an automatic test at some point, but what I meant here is [this](https://github.com/EdgeLab-FHDO/Edge-Diagnostic-Platform/wiki/Running-Scenarios). Try to do it yourself first, then we have a call about the results.

**Comment by zero212** [comment]
When updating node or client, should we update the values or should we replace the whole object with a new one?
I don't see any differences in either method, to be honest...

**Comment by alyhasansakr** [comment]
No difference.

**Comment by zero212** [comment]
What kind of scenario do we need? I'm done with debugging/cleaning up the code in match making and SBMM. 
Just whatever scenario i want? As long as it demonstrates the SBMM functions and match maker character (disconnect, history after etc) is okay?

**Comment by alyhasansakr** [comment]
Yes, whatever you want as long as it shows the different parts of the implementation.

**Comment by zero212** [comment]
Future work:
- Weights should be dynamic
- make a logger for debugging (using a toggle) since we are shifting toward read() function in masterInput 
- Implement visibility, This rather connected to "input" calculate scores, take the input and somehow "blur" the info by introducing some deviant (multiply by 5-30%, so we get the estimation)

**Comment by zero212** [comment]
MERRY CHRISTMAS YOU ALLL. ENJOY YOUR HOLIDAY OVER THERE
Stay safe and have fun
@alyhasansakr  @juan-castrillon  @JeanValjean17  @theodorelu27
